Remove Hyper-V as a hard dependency

aserdean opened the pull request:
Remove Hyper-V as a hard dependency

This patch includes an updated version of the OVS installer on Windows.

This patch also provides updates to the ansible playbooks to accommodate
for a non Hyper-V enabled host.

This patch also includes fixes and updates to the ansible playbooks.

Signed-off-by: Alin Gabriel Serdean <aserdean@ovn.org>
Signed-off-by: Alin-Gheorghe Balutoiu <alinbalutoiu@gmail.com>
Co-authored-by: Alin-Gheorghe Balutoiu <alinbalutoiu@gmail.com>

alinbalutoiu commented on contrib/roles/linux/ovn-kubernetes/vars/ubuntu.yml:
```diff
@@ -2,9 +2,9 @@
+  git_url:  https://github.com/aserdean/ovn-kubernetes
   build_path: /tmp
-  branch: master
+  branch: no_hyperv
```
This is just for testing, should be changed back to the original repo and master branch for final PR.

aserdean replied:
Yup. Everything is very `alpha` here :)

alinbalutoiu commented on contrib/roles/windows/docker/tasks/requirements-16299.yaml:
```diff
@@ -8,9 +8,9 @@
-    - Hyper-V-PowerShell # TODO
+    #- Hyper-V # TODO
+    #- RSAT-Hyper-V-Tools # TODO
+    #- Hyper-V-PowerShell # TODO
```
Just remove those lines instead of commenting them.

aserdean replied:
Hmmm. Should we bother we Hyper-V features via some variable?

alinbalutoiu commented on contrib/roles/windows/docker/tasks/requirements-16299.yaml:
```diff
@@ -8,9 +8,9 @@
+    #- RSAT-Hyper-V-Tools # TODO
+    #- Hyper-V-PowerShell # TODO
 
 - name: Windows 1709 | Expect reboot_required to false
```
Rename the filename to ``requirements.yaml`` and put ``Windows Server`` instead of ``Windows 1709`` as those requirements will be for all versions of Windows.

aserdean replied:
Sure

alinbalutoiu commented on contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml:
```diff
@@ -12,42 +12,28 @@
-      # Import-Module ./hns.psm1
+      #TODO figure why they don't get automatically included
+      Import-Module "C:\Windows\System32\WindowsPowerShell\v1.0\Modules\HostNetworkingService\HostNetworkingService.psm1"
+      Get-HNSNetwork
```
Is this line required?

aserdean replied:
Nope. I will remove it in the update

alinbalutoiu commented on contrib/roles/windows/kubernetes/tasks/setup_ovs_docker.yml:
```diff
@@ -12,42 +12,28 @@
-      # Remove-NetIPAddress -ifAlias br-ex -Confirm:$false
-      # New-NetIPAddress -ifAlias br-ex -IPAddress {{ host_internal_ip }} -DefaultGateway {{ interface_default_gateway }} -PrefixLength {{ interface_prefix_length }}
-      # Set-DnsClientServerAddress -InterfaceAlias br-ex -ServerAddresses ("8.8.8.8")
+      # TODO: Carry over MTU to the management interface
```
This should be implemented in order to have this working on clouds that have MTU requirements (such as GCE).

aserdean replied:
Yup. Should I split this in a different PR?

alinbalutoiu commented on go-controller/pkg/ovn/management-port.go:
```diff
@@ -19,7 +19,7 @@ const (
 	routerIP, interfaceName, interfaceIP string) error {
 	// Up the interface.
-	args := []string{"Enable-NetAdapter", fmt.Sprintf("%s", interfaceName)}
+	args := []string{"Enable-NetAdapter", "-IncludeHidden", fmt.Sprintf("%s", interfaceName)}
```
Can you put the changes of the go files in a separate commit?

aserdean replied:
Ofc

aserdean replied:
https://github.com/openvswitch/ovn-kubernetes/pull/372
sent

alinbalutoiu commented on go-controller/pkg/util/ovs.go:
```diff
@@ -14,7 +14,7 @@ import (
 
 const (
-	ovsCommandTimeout = 5
+	ovsCommandTimeout = 15
```
I think a comment would be good here explaining why the timeout has been increased that much.

aserdean replied:
Sure.

alinbalutoiu commented:
Added some comments to the PR.
I tested on Windows Server 1803 (buildnumber 17134 and all the updates installed) and encountered the following issues:
- it looks like HNS doesn't report that the extension is enabled even though it is actually enabled and running
```
@{Id=583cc151-73ec-4a6a-8b47-578297ad7623; IsEnabled=False; Name=Cloudbase Open vSwitch Extension},
```
this makes the call to Get-OVSEnabledHNSNetworks to fail since it won't find any HNS network that has the extension enabled. Changed the definition of Get-OVSEnabledHNSNetworks to return the HNS network without checking if extension is enabled allowed it to work. Note: it seems that not even restart fixes the issue and it still shows extension is disabled after restart.

- can you also change this check? https://github.com/openvswitch/ovn-kubernetes/blob/master/contrib/roles/windows/kubernetes/tasks/main.yml#L41
otherwise it will always run setup_ovs_docker on every playbook run. You should check if the bridge is created and skip the task, or even ``when: transparent_inspect.rc != 0`` should work fine as well

aserdean commented:
@rajatchopra , @alinbalutoiu I updated the PR and tested on 1709 and also 1803. Can you please test it out?

rajatchopra commented:
After a marathon session, Alin helped me get a cluster up and running with this PR. Thanks @aserdean .
This PR lgtm. Ready for merge, pending the branch name change.

The other issue of finding the default interface for docker network setup will dealt with separately.

alinbalutoiu commented on contrib/roles/windows/kubernetes/tasks/main.yml:
```diff
@@ -38,7 +38,7 @@
 - name: Kubernetes | OVS Docker network setup
   include_tasks: ./setup_ovs_docker.yml
-  when: interface_name != "br-ex"
+  when: transparent_inspect.rc != 0
```
This check should be improved. I think the following should work:
```
- name: Kubernetes | Check if system-id is set
  win_shell: ovs-vsctl get Open_vSwitch . external_ids:system-id 
  register: ovs_system_id
  ignore_errors: true
  changed_when: false

- name: Kubernetes | OVS Docker network setup
  include_tasks: ./setup_ovs_docker.yml
  when: ovs_system_id.rc != 0
```

aserdean replied:
It still won't be sufficient to be honest.
I'll try to add some powershell commandlets to the OVS installer to improve the network creation/check/removal and send out a different PR for it.

alinbalutoiu commented on contrib/roles/windows/kubernetes/tasks/run_minion_init.yml:
```diff
@@ -19,6 +19,36 @@
+- name: Kubelet | Create ovn-kubernetes-node service
+  win_service:
+    name: ovn-kubernetes-node
+    path: '"{{ install_info.install_path }}\\servicewrapper.exe" ovn-kubernetes-node {{ install_info.install_path }}\\ovnkube.exe --k8s-kubeconfig={{ install_info.install_path }}\\kubeconfig.yaml --k8s-apiserver http://{{ kubernetes_info.MASTER_IP }}:8080 --init-node {{ ansible_hostname|lower }} --k8s-token {{TOKEN}} --nb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6641" --sb-address "tcp://{{ ...
```
Please add the following argument ``--encap-ip={{ host_internal_ip }}`` to ovnkube.

aserdean replied:
I don't really like this approach because the adapter may change(i.e. DHCP server) . We need to improve the IP detection in the end IMO.
I will add it as a workaround and open an issue for the IP detection.

alinbalutoiu commented on contrib/roles/windows/kubernetes/tasks/run_minion_init.yml:
```diff
@@ -27,9 +57,8 @@
-      {{ install_info.install_path }}\\ovnkube.exe --k8s-kubeconfig={{ install_info.install_path }}\\kubeconfig.yaml --k8s-apiserver http://{{ kubernetes_info.MASTER_IP }}:8080 --init-node {{ ansible_hostname|lower }} --k8s-token {{TOKEN}} --nb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6641" --sb-address "tcp://{{ kubernetes_info.MASTER_IP }}:6642" --cluster-subnet {{ kubernetes_info.CLUSTER ...
+      Start-Service ovn-kubernetes-node
+      sleep 15
       ovs-vsctl set open_vswitch . external_ids:ovn-encap-ip="{{ host_public_ip }}"
```
The sleep and setting ovn-encap-ip can be removed if ``--encap-ip={{ host_internal_ip }}`` is passed to ovnkube.

aserdean replied:
Same as above. I will add it as a workaround and open an issue for the IP detection.

aserdean commented:
@alinbalutoiu, @rajatchopra thanks a lot for trying and testing this PR.
I addressed all of the comments and updated the commit message, can you please take a final look?
Also thanks a lot @ionutbalutoiu for hosting the new installer until we get a signed version.

rajatchopra commented on contrib/roles/windows/kubernetes/vars/windows.yml:
```diff
@@ -13,7 +13,7 @@ kubernetes_info:
   K8S_DNS_SERVICE_IP: "{{K8S_DNS_SERVICE_IP | default('10.0.9.10')}}"
   K8S_API_SERVICE_IP: "{{K8S_API_SERVICE_IP | default('10.0.9.1')}}"
-  infracontainername_1709: "ovnkubernetes/pause"
+  infracontainername: "ovnkubernetes/pause"
```
will this eventually mean docker.io/ovnkubernetes/pause?
Let us discuss the final place for the community repository.

rajatchopra commented:
/lgtm
tested ok.
